### ISSUE TITLE
More Documentation Needed to Avoid RunTime errors

### ISSUE THREAD

**Issue by elibroftw** [home-assistant-libs/pychromecast#560]
Suppose I have a cast `cast` and have already called `cast.wait()`
Subsequent `cast.wait()`'s result in RunTimeError's if the thread has already been run and if these errors do occur, then the `self.status_event.wait(timeout=timeout)` never happens. This is what I suspect indirectly causes my NotConnected errors when I try to set_volume afterwards.

I will test if the below fixes the problem for me, but before updating pychromecast by 2 years worth of code, no RunTimeError's were being raised by `cast.wait()`. (EDIT: it does not fix the underlying problem).
```py
def cast_wait():
    try:
        cast.wait(timeout=WAIT_TIMEOUT)
    except AttributeError:
        pass
    except RuntimeError:
        cast.status_event.wait(WAIT_TIMEOUT)
```

**Comment by elibroftw** [comment]
I suspect the socket_client was modified in the last 2 years and removed some threading checks.

**Comment by emontnemery** [comment]
Please be a bit more precise, what are you trying to do and what goes wrong?

**Comment by emontnemery** [comment]
It should be perfectly fine to call `Chromecast.wait` multiple times.

Please share a step-by-step description, or even better, Python code, explaining:
- What you are trying to do, **with details**
- What did you expect to happen
- What happened instead

**Comment by elibroftw** [comment]
This function https://github.com/elibroftw/music-caster/blob/master/src/music_caster.py#L1461

The cast_wait call from this branch https://github.com/elibroftw/music-caster/blob/master/src/music_caster.py#L1485 used to be just a with suppress(RunTimeErrror): cast.wait() but then the set volume function resulted in an error. 

I cannot reproduce the conditions to cause the error because the error is caused when the internal thread gets exhausted and I do not know when or in what cases that happens.

**Comment by emontnemery** [comment]
Thanks.

When you write "the internal thread gets exhaused", do you mean the `SocketClient`'s worker thread has died?
If that has happened, all hope is lost and you will need to create a new `Chromecast` object.

That's not expected to happen, it should run forever and automatically reconnect as needed, unless the `Chromecast` object was created with limited retries (not the default).

This PR https://github.com/home-assistant-libs/pychromecast/pull/390 added some extra prints to log unexpected exceptions, please upgrade to a version including those.

**Comment by elibroftw** [comment]
Well yes it has to do with the socket client. If the socket client is supposed to run forever, I suggest you recreate the thread rather than make the client recreate the cast.

Right now I'm bypassing the wait function by calling status_event.wait() myself. I've only been running my app for a day but I do believe this has fixed the errors. So if that thread is never supposed to stop or die, then you should at least catch the runtimeerror and restart the thread rather than have the client recreate the cast when all the inputs are the same.

**Comment by emontnemery** [comment]
Your bypass won't solve anything, the `Chromecast` object will need to be recreated if the worker died, it's no longer functional.
Please upgrade pychromecast, it will then be clear why the thread dies and we can get that fixed.

> It could be the unlimited retires thing, so I'll see if it had to do with my discovery.

Right, you need to make sure `SocketClient` has `tries` set to 0 or `None`.

**Comment by elibroftw** [comment]
I'm already using the latest version.

**Comment by emontnemery** [comment]
Can't you modify your production executable to store Python logs in a file or something like that?
pychromecast uses `logging` exclusively, so you ndon't need to worry about redirecting stdout or stderr.

It should be possible to modify this:
```py
        try:
            while not self.stop.is_set():
                if self.run_once(timeout=POLL_TIME_BLOCKING) == 1:
                    break
        except Exception:  # pylint: disable=broad-except
            self.logger.exception(
                ("[%s(%s):%s] Unhandled exception in worker thread"),
                self.fn or "",
                self.host,
                self.port,
            )
            raise
```
To:
```py
        while not self.stop.is_set():
            try:
                if self.run_once(timeout=POLL_TIME_BLOCKING) == 1:
                    break
            except Exception:  # pylint: disable=broad-except
                self.logger.exception(
                    ("[%s(%s):%s] Unhandled exception in worker thread"),
                    self.fn or "",
                    self.host,
                    self.port,
                )
```

I'm a bit worried about the SocketClient spinning out of control though.

**Comment by elibroftw** [comment]
Did you say that retries is by default unlimited? In that case, how can it still be the stocketclient crashing?

**Comment by elibroftw** [comment]
And the error messages I was talking about only happened when I was testing out `cc.connect()` after `cc.wait()` which I don't do in my app.

**Comment by elibroftw** [comment]
Is there anyway to set the logger to a file at runtime?

**Comment by emontnemery** [comment]
> Did you say that retries is by default unlimited? In that case, how can it still be the socketclient crashing?

SocketClient will catch expected exceptions and if they happen attempt to reconnect to the remote device indefinitely, unless you have set `tries` to a non-zero integer.
If something unexpected goes wrong, the thread log the unexpected and then terminate.

> the error messages I was talking about only happened when I was testing out cc.connect() after cc.wait()

Don't do that. You should all connect only if you don't want the `SocketClient` to start its own thread, this is perfectly clear from the docstrings for `connect` and `wait`.

> Is there anyway to set the socketclient logger to my own log handler?

Please refer to the Python documentation or SE.

**Comment by elibroftw** [comment]
I set the handler of the logger to my own so let me do some more testing.

**Comment by elibroftw** [comment]
2021-11-17 11:47:46,693 ERROR (597): [Office speaker(192.168.1.6):8009] Error reading from socket.
2021-11-17 11:47:46,693 WARNING (650): [Office speaker(192.168.1.6):8009] Error communicating with socket, resetting connection
2021-11-17 11:47:59,694 ERROR (528): [Office speaker(192.168.1.6):8009] Unhandled exception in worker thread
Traceback (most recent call last):
  File "pychromecast\socket_client.py", line 525, in run
  File "pychromecast\socket_client.py", line 548, in run_once
  File "pychromecast\socket_client.py", line 677, in _check_connection
  File "pychromecast\socket_client.py", line 294, in initialize_connection
  File "pychromecast\dial.py", line 37, in get_host_from_service
  File "zeroconf\_core.py", line 523, in get_service_info
  File "zeroconf\_services\info.py", line 458, in request
  File "zeroconf\_utils\asyncio.py", line 95, in run_coro_with_timeout
  File "concurrent\futures\_base.py", line 446, in result
concurrent.futures._base.TimeoutError
2021-11-17 12:49:58,005 INFO (1976): v

**Comment by elibroftw** [comment]
This error happened after not interacting with the chromecast for ~30+ minutes.

**Comment by emontnemery** [comment]
Ouch, I think that's a bug in pychromecast, it should return `None` and not throw if the service can't be resolved.

**Comment by elibroftw** [comment]
But that won't fix the connection issue. How can I ensure that the Chromecast will get connected to after being idle for a while?

**Comment by emontnemery** [comment]
The error in pychromecast literally happens when pychromecast is trying to reconnect, your cast device must have dropped off from the network or rebooted.

**Comment by elibroftw** [comment]
No it didn't. I can connect to it on another instance of my app (the Python version) but the production version can't connect. I've had problems with zeroconf in the past so this isn't completely new.

**Comment by emontnemery** [comment]
The stack trace you show shows pychromecast reconnecting, and that failed because zeroconf unexectedly throws.
I'm not saying your app or chromecast is at fault, just explaining what's going on.

Can you try to add this patch to pychromecast:
```diff
diff --git a/pychromecast/dial.py b/pychromecast/dial.py
index 28dc05c..77b19df 100644
--- a/pychromecast/dial.py
+++ b/pychromecast/dial.py
@@ -1,6 +1,7 @@
 """
 Implements the DIAL-protocol to communicate with the Chromecast
 """
+import asyncio
 from collections import namedtuple
 import json
 import logging
@@ -42,7 +43,7 @@ def get_host_from_service(service, zconf):
                 service,
                 service_info,
             )
-    except IOError:
+    except (IOError, asyncio.TimeoutError):
         pass
     return _get_host_from_zc_service_info(service_info) + (service_info,)

diff --git a/pychromecast/discovery.py b/pychromecast/discovery.py
index 0143bd1..0b907b7 100644
--- a/pychromecast/discovery.py
+++ b/pychromecast/discovery.py
@@ -1,5 +1,6 @@
 """Discovers Chromecasts on the network using mDNS/zeroconf."""
 import abc
+import asyncio
 import functools
 import itertools
 import logging
@@ -150,7 +151,7 @@ class ZeroConfListener:
         while service is None and tries < 4:
             try:
                 service = zconf.get_service_info(typ, name)
-            except IOError:
+            except (IOError, asyncio.TimeoutError):
                 # If the zeroconf fails to receive the necessary data we abort
                 # adding the service
                 break
```

**Comment by elibroftw** [comment]
I'm using it right.

This is good for preventing an error from being raised but the consequential error of not being able to use the Chromecast after wait() still exists.

**Comment by emontnemery** [comment]
I chatted a bit with one of the authors of python-zeroconf.

It's quite unexpected to have an `asyncio.TimeoutError` happen.
Could you try to get some more info on what's triggering it?

For example:
- Does it happen when CPU-load is very high?
- Do you use asyncio in your app, directly or indirectly via some library?
- Does it happen of you hard-reboot a chromecast your app is connected to, e.g. pull the USB-cable
- Does your app end up creating multiple zeroconf instances in parallel? This will happen if you repeatedly call pychromcast discovery APIs without passing a shared zeroconf or without stopping previous zeroconf instances before the next call

**Comment by emontnemery** [comment]
> the consequential error of not being able to use the Chromecast after wait() still exists.

I really don't see that happening.
The patch above will prevent the pychromecast thread from exiting, and you can call wait() repeatedly with no issue.

**Comment by emontnemery** [comment]
There is however a big issue in your code:
https://github.com/elibroftw/music-caster/blob/941effe7487473ae8803134c8714b703ca69731c/src/music_caster.py#L1011-L1024

This won't work, you need to keep the browser alive during the lifetime of the `Chromecast` object, or pychromecast won't be able to reconnect.
There's also no need to kick off the discovery in its own thread, that will only create problems.

Take a look at `examples/discovery_example.py`, it will start the discovery - which runs in its own thread automatically - and gets callbacks through an implementation of `pychromecast.discovery.AbstractCastListener`.

When the `add_cast` callback is called - from the zeroconf browser's thread - you can create a `Chromecast` object by calling `pychromecast.get_chromecast_from_cast_info`, passing the `service`.

So what you should do is:
- Create a `pychromecast.discovery.CastBrowser` when your app starts
- When the wanted chromecast is available - as notified by the `add_cast` callback, connect to it
- Kill the browser only when your app exits

**Comment by elibroftw** [comment]
Oh okay, so the bug is just me stopping the discovery needlessly then?

**Comment by elibroftw** [comment]
I can't do exactly what you recommend because I have a function called "Refresh Devices" in case the user decides to turn on their TV for example. But I will remove the stop_discovery since that is probably what's causing the errors then?

**Comment by elibroftw** [comment]
Okay I cleaned up that function now so that it doesn't double thread.

**Comment by elibroftw** [comment]
I think the issue was with me stopping the discovery so I'll close this issues prematurely.

**Comment by bdraco** [comment]
`zeroconf` 0.37.0 is now available.  It should provide some better error reporting for cases like this

https://github.com/jstasiak/python-zeroconf/releases

**Comment by emontnemery** [comment]
> I can't do exactly what you recommend because I have a Refresh Devices feature in case the user decides to turn on their TV for example. But I will remove the stop_discovery since that is probably what's causing the errors then?

If you do as I suggested, your "refresh devices" routine really just needs to show the current list of devices discovered, there's no need - and you should not - call `pychromecast.get_chromecasts` again.
You simply need to check the browser's `devices` list, check `list_devices` in `discovery_example.py`.

Note that `CastBrowser.devices` is updated by the `CastBrowser`'s thread, so you need to copy it before iterating over it:
```py
for device in list(browser.devices):
   ...
```

**Comment by elibroftw** [comment]
So it'd be prudent to just remove checker since pychromecast will simply already able to detect new devices without requiring input from my part?

**Comment by elibroftw** [comment]
Okay let me ask for clarification for one more point.

Suppose a Chromecast was disconnected after the cast browser detected it. Would the list be automatically up to date without the Chromecast that just got disconnected from the network?

**Comment by emontnemery** [comment]
> Suppose a Chromecast was disconnected after the cast browser detected it. Would the list be automatically up to date without the Chromecast that just got disconnected from the network?

If the Chromecast is rebooted or turned off, it will disappear immediately from the list because the Chromecast will remove its mDNS service.
If the Chromecast unexpectedly disappears because it crashes or the power cable is yanked it will disappear from the list in a couple of minutes (when the mDNS service expires).

**Comment by bdraco** [comment]
> If the Chromecast unexpectedly disappears because it crashes or the power cable is yanked it will disappear from the list in a couple of minutes (when the mDNS service expires).

Usually < 120s

**Comment by elibroftw** [comment]
Okay I've changed my code to use a Listener now.
- When is update_cast() important when using a custom listener?

**Comment by elibroftw** [comment]
And now no chromecasts are detected when running my app as a script but it works fine in the latest release of it...
Anyway to debug why this issue is happening? 
 remember this was the issue due to zeronconf that forced me to use an older version of pychromecast for more than 2 years.

**Comment by emontnemery** [comment]
> When is update_cast() important when using a custom listener

You don't really need to worry about it TBH.

> no chromecasts are detected when running my app as a script on my laptop but it works fine in the latest compiled version and on my Desktop

Please have a look at the example scripts, they all allow configuring log level with a separate setting for zeroconf.

**Comment by elibroftw** [comment]
I think when this error was patched, now what shows up is a generic "pychromecast.error.PyChromecastError" without any message.
Logs say this:
```
2022-03-05 16:32:36,391 WARNING (672): [Musik(192.168.0.3):8009] Heartbeat timeout, resetting connection
```
